File: python/problems/leetcode/0101_symmetric_tree/solution.py
"""LeetCode 101: Symmetric Tree."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def mirror(a: Optional[TreeNode], b: Optional[TreeNode]) -> bool:
            if a is None or b is None:
                return a is None and b is None
            return (
                a.val == b.val
                and mirror(a.left, b.right)
                and mirror(a.right, b.left)
            )

        return root is None or mirror(root.left, root.right)


if __name__ == "__main__":
    # [1,2,2,3,4,4,3]
    root = TreeNode(
        1,
        TreeNode(2, TreeNode(3), TreeNode(4)),
        TreeNode(2, TreeNode(4), TreeNode(3)),
    )
    assert Solution().isSymmetric(root)
    # [1,2,2,null,3,null,3]
    bad = TreeNode(
        1, TreeNode(2, None, TreeNode(3)), TreeNode(2, None, TreeNode(3))
    )
    assert not Solution().isSymmetric(bad)
    print("0101 OK")

File: cpp/problems/leetcode/0101_symmetric_tree/solution.cpp
// LeetCode 101
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        return mirror(root->left, root->right);
    }

private:
    bool mirror(TreeNode* a, TreeNode* b) {
        if (!a || !b) return a == b;
        return a->val == b->val && mirror(a->left, b->right) &&
               mirror(a->right, b->left);
    }
};

int main() {
    TreeNode l3a(3), l4a(4), l4b(4), l3b(3);
    TreeNode l2a(2, &l3a, &l4a);
    TreeNode l2b(2, &l4b, &l3b);
    TreeNode root(1, &l2a, &l2b);
    assert(Solution().isSymmetric(&root));

    TreeNode r3a(3), r3b(3);
    TreeNode r2a(2, nullptr, &r3a);
    TreeNode r2b(2, nullptr, &r3b);
    TreeNode bad(1, &r2a, &r2b);
    assert(!Solution().isSymmetric(&bad));
    cout << "0101 OK" << endl;
    return 0;
}
